### PR TITLE
[Issue-64] Delete Batch Pool operation via WSM

### LINKED ISSUE #64: Use WSM to delete batch pools
Deleting batch pools from WSM is working in progress. 

AC:
 - Refactor the Terra Batch Pool manager to use the WSM endpoint.
 - Add the delete batch pool operation to the WSM API client

### PR DESCRIPTION
**Pull request by giventocode**
Closes #64 

This PR includes:
- Batch Pool deletion is via WSM when TES is running in Terra
- New WSM API in the client to delete the batch pool via the WSM resource id.

### CODE REVIEW

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs [hunk 1]
```diff
@@ -124,19 +124,17 @@ public async Task DeleteBatchPoolAsync(string poolId, CancellationToken cancella
...
 
-                await batchClient.PoolOperations.DeletePoolAsync(poolId, cancellationToken: cancellationToken);
+                var wsmResourceId = await GetWsmResourceIdFromBatchPoolMetadataAsync(poolId, cancellationToken);
 
-                logger.LogInformation(
-                    $"Successfully deleted pool with the id/name:{poolId}");
+                await terraWsmApiClient.DeleteBatchPoolAsync(Guid.Parse(terraOptions.WorkspaceId), wsmResourceId);
 
+                logger.LogInformation(
+                    $"Successfully deleted pool with the id/name via WSM:{poolId}");
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:137]: could add a space after WSM:

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs [hunk 2]
```diff
@@ -146,6 +144,28 @@ public async Task DeleteBatchPoolAsync(string poolId, CancellationToken cancella
...
 
+        private async Task<Guid> GetWsmResourceIdFromBatchPoolMetadataAsync(string poolId, CancellationToken cancellationToken)
+        {
+            var batchClient = CreateBatchClientFromOptions();
+
+            var pool = await batchClient.PoolOperations.GetPoolAsync(poolId, cancellationToken: cancellationToken);
+
+            if (pool is null)
+            {
+                throw new InvalidOperationException($"The batch pool was not found. Pool id:{poolId}");
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:155]: capitalize "Batch", and capialize "ID" and space after "ID: "

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs [hunk 3]
```diff
@@ -146,6 +144,28 @@ public async Task DeleteBatchPoolAsync(string poolId, CancellationToken cancella
...
+            }
+
+            var metadataItem = pool.Metadata.SingleOrDefault(m => m.Name.Equals(TerraResourceIdMetadataKey));
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:158]: should add OrdinalIgnoreCase to this

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs [hunk 4]
```diff
@@ -146,6 +144,28 @@ public async Task DeleteBatchPoolAsync(string poolId, CancellationToken cancella
...
+
+            if (metadataItem is null || string.IsNullOrEmpty(metadataItem.Value))
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:160]: if (string.IsNullOrEmpty(metadataItem?.Value))

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs [hunk 5]
```diff
@@ -146,6 +144,28 @@ public async Task DeleteBatchPoolAsync(string poolId, CancellationToken cancella
...
+            {
+                throw new InvalidOperationException("The WSM resource id was not found in the pool's metadata.");
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:162]: ID

File: src/TesApi.Web/Management/Clients/TerraWsmApiClient.cs [hunk 6]
```diff
@@ -89,24 +90,77 @@ public virtual async Task<ApiCreateBatchPoolResponse> CreateBatchPool(Guid works
...
-                Logger.LogError(ex, $"Failed to create a batch pool using wsm. Response content:{responseContent}");
+                Logger.LogInformation($"Deleting the batch pool using WSM for workspace:{workspaceId} WSM resource id:{wsmBatchPoolResourceId}");
+
+                response =
+                    await HttpSendRequestWithRetryPolicyAsync(() => new HttpRequestMessage(HttpMethod.Delete, uri),
+                        setAuthorizationHeader: true);
+
+                response.EnsureSuccessStatusCode();
+
+                Logger.LogInformation($"Successfully deleted  batch pool, WSM resource id: {wsmBatchPoolResourceId} using WSM for workspace:{workspaceId}");
```
Comment (MattMcL4475) [inline src/TesApi.Web/Management/Clients/TerraWsmApiClient.cs:122]: extra space between deleted and batch (Batch), ID

**Review by MattMcL4475 — APPROVED** [review]

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs (hunk 3 above)
Comment (giventocode) [reply src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:158]: `TerraResourceIdMetadataKey` is a constant and used in when setting and reading. The case should be exact, no need to assume other.

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs (hunk 4 above)
Comment (giventocode) [reply src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:160]: No need to check for null on the second part of `or` the statement - short-circuit evaluation in C#

Comment (giventocode) [reply src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:160]: Never mind, I miss read the suggestion. Makes sense. I agree is better.

File: src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs (hunk 3 above)
Comment (MattMcL4475) [reply src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:158]: it's faster to use `StringComparison.OrdinalIgnoreCase` if we can! :-)

Comment (BMurri) [reply src/TesApi.Web/Management/Batch/TerraBatchPoolManager.cs:158]: It's actually both faster and more reliable to add either `StringComparison.OrdinalIgnoreCase` or `StringComparison.Ordinal` than to leave it bare. Leaving it bare causes locale rules based on current culture settings to both slow things down and to sometimes produce different results.

**Pull request merged**